balancer/weightedroundrobin/balancer.go

```diff
@@ -527,21 +592,35 @@ func (w *weightedSubConn) updateConnectivityState(cs connectivity.State) connect
 
 // weight returns the current effective weight of the subconn, taking into
 // account the parameters.  Returns 0 for blacked out or expired data, which
-// will cause the backend weight to be treated as the mean of the weights of
-// the other backends.
-func (w *weightedSubConn) weight(now time.Time, weightExpirationPeriod, blackoutPeriod time.Duration) float64 {
+// will cause the backend weight to be treated as the mean of the weights of the
+// other backends. If forScheduler is set to true, this function will emit
+// metrics through the mtrics registry.
+func (w *weightedSubConn) weight(now time.Time, weightExpirationPeriod, blackoutPeriod time.Duration, forScheduler bool) (weight float64) {
 	w.mu.Lock()
 	defer w.mu.Unlock()
+
+	weight = 0
+	if forScheduler {
+		defer endpointWeightsHandle.Record(w.metricsRecorder, weight, []string{w.target, w.locality}...)
+	}
+
 	// If the most recent update was longer ago than the expiration period,
 	// reset nonEmptySince so that we apply the blackout period again if we
 	// start getting data again in the future, and return 0.
 	if now.Sub(w.lastUpdated) >= weightExpirationPeriod {
+		if forScheduler {
+			endpointWeightStaleHandle.Record(w.metricsRecorder, 1, []string{w.target, w.locality}...)
+		}
 		w.nonEmptySince = time.Time{}
-		return 0
+		return
```

Comment:
Please (almost?) never use bare returns from functions that return values.  It's too hard to read.  Use `return weight` or, even better, keep as `return 0`.

Reply:
I'll return weight so reader doesn't need to keep track of two pieces of information (named return value which gets logged and the actual return value).

Reply:
IMO this is worse because now you have to go backtrack and figure out what `weight` might be while reading the code.  If it just says `return 0` then it's obvious.

Note that named return values are set upon returning.  I.e.

```go
func f() (x int) {
	x = 5
	defer func() {
		fmt.Println(x)
	}()

	return 10
}
```

Will print `10`.

Reply:
Ah ok I didn't realize named return values were set upon returning. Switched to incorporate all your suggestions.